java/core/src/main/java/com/google/protobuf/CodedInputStream.java

```diff
@@ -825,13 +825,9 @@ public String readString() throws IOException {
     public String readStringRequireUtf8() throws IOException {
       final int size = readRawVarint32();
       if (size > 0 && size <= (limit - pos)) {
-        // TODO(martinrb): We could save a pass by validating while decoding.
```

Comment:
Can we guard this change by a static boolean? That way it would be easier to roll back if there are performance regressions.

Reply:
Added a constant that reads a system property. It means javac won't remove the branches but JVM will easily do so, but let me know if you'd prefer a compiler constant.